lib/js/jsbigints.nim

```diff
@@ -0,0 +1,137 @@
+
+func newBigInt*(integer: SomeInteger): JsBigInt {.importjs: "BigInt(#)".}
+  ## Constructor for `JsBigInt`.
+
+func big*(integer: cstring): JsBigInt {.importjs: "BigInt(#)".} =
+  ## Constructor for `JsBigInt`.
+  runnableExamples:
+    doAssert big"-1" == big"1" - big"2"
+
+func toLocaleString*(this: JsBigInt): cstring {.importjs: "#.$1()".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toLocaleString
+
+func toLocaleString*(this: JsBigInt; locales: cstring): cstring {.importjs: "#.$1(#)".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toLocaleString
+
+func toLocaleString*(this: JsBigInt; locales: openArray[cstring]): cstring {.importjs: "#.$1(#)".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toLocaleString
+
+func toString*(this: JsBigInt; radix: int): cstring {.importjs: "#.$1(#)".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toString
+
+func toString*(this: JsBigInt): cstring {.importjs: "#.toString()".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/toString
+
+func `$`*(this: JsBigInt): string = $toString(this)
+  ## Return a string representation of `JsBigInt`.
+
+func asIntN*(width: int; bigInteger: JsBigInt): int {.importjs: "BigInt.$1(#, #)".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/asIntN
+
+func asUintN*(width: int; bigInteger: JsBigInt): int {.importjs: "BigInt.$1(#, #)".}
+  ## https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt/asUintN
+
+func `+`*(x, y: JsBigInt): JsBigInt {.importjs: "(# $1 #)".}
+
+func `-`*(x, y: JsBigInt): JsBigInt {.importjs: "(# $1 #)".}
+
+func `*`*(x, y: JsBigInt): JsBigInt {.importjs: "(# $1 #)".}
+
+func `div`*(x, y: JsBigInt): JsBigInt {.importjs: "(# / #)".}
+   ## Same as `div` but for `JsBigInt`.
+
+func `mod`*(x, y: JsBigInt): JsBigInt {.importjs: "(# % #)".}
+  ## Same as `mod` but for `JsBigInt` (uses JavaScript `BigInt() % BigInt()`).
+
+func `+=`*(x, y: JsBigInt) {.importjs: "(# $1 #)".}
+
+func `-=`*(x, y: JsBigInt) {.importjs: "(# $1 #)".}
+
+func `*=`*(x, y: JsBigInt) {.importjs: "(# $1 #)".}
+
+func `div=`*(x, y: JsBigInt) {.importjs: "(# /= #)".}
+
+func `mod=`*(x, y: JsBigInt) {.importjs: "(# $1 #)".}
+
+func `inc`*(x: JsBigInt) {.importjs: "(++#)".}
+  ## Same as `inc` but for `JsBigInt` (uses JavaScript `++BigInt()`).
+
+func `dec`*(x: JsBigInt) {.importjs: "(--#)".}
+  ## Same as `dec` but for `JsBigInt` (uses JavaScript `--BigInt()`).
+
+func `<`*(x, y: JsBigInt): bool {.importjs: "(# $1 #)".}
+
+func `<=`*(x, y: JsBigInt): bool {.importjs: "(# $1 #)".}
+
+func `==`*(x, y: JsBigInt): bool {.importjs: "(# === #)".}
+
+func `**`*(x, y: JsBigInt): JsBigInt {.importjs: "((#) $1 #)".}
```

Comment:
why `((#) $1 #)` here but in other cases you use `(# $1 #)` ?

Reply:
for `func(12 + 4,  4)` or similar, I can't remember though

Reply:
see https://github.com/xflywind/Nim/blob/8f9aac66f459d5031ccfa33562617a4b2900a035/tests/js/tjsffi.nim#L362

Reply:
https://github.com/nim-lang/Nim/blob/devel/lib/js/jsffi.nim#L182

```javascript
$ node
Welcome to Node.js v15.4.0.
> 4 ** 2
16
> -4 ** 2
-4 ** 2
^^^^^
Uncaught:
SyntaxError: Unary operator used immediately before exponentiation expression. 
Parenthesis must be used to disambiguate operator precedence.

>
```

Reply:
Please test this special case or add it to runnableExamples

Reply:
The JS standard says it *must* be like that anyways, yes theres an assert with it.

Reply:
see full explanation here: https://github.com/nim-lang/Nim/pull/16409#issuecomment-760550812

> for func(big"12" + big"4", big"4") or similar, I can't remember though

not this

> see https://github.com/xflywind/Nim/blob/8f9aac66f459d5031ccfa33562617a4b2900a035/tests/js/tjsffi.nim#L362

unfortunately, this doesn't exercise the critical test involving a unary operator, we need to add a regression test